internal/checker/checker.go

```diff
@@ -20903,8 +20903,21 @@ func (c *Checker) getTypeFromTypeNode(node *ast.Node) *Type {
 
 func (c *Checker) getTypeFromTypeNodeWorker(node *ast.Node) *Type {
 	switch node.Kind {
-	case ast.KindAnyKeyword:
+	case ast.KindAnyKeyword, ast.KindJSDocAllType:
 		return c.anyType
+	case ast.KindJSDocNonNullableType:
+		return c.getTypeFromTypeNode(node.AsJSDocNonNullableType().Type)
+	case ast.KindJSDocNullableType:
+		t := c.getTypeFromTypeNode(node.AsJSDocNullableType().Type)
+		if c.strictNullChecks {
+			return c.getNullableType(t, TypeFlagsNull)
```

Comment:
No undefined?

Reply:
we type `?T` with the Closure-derived type `T | null` but the TS error suggests the Flow-derived `T | null | undefined`. That's because I thought that in a JS file we should give the Closure type, but that in a TS file, we were likely to encounter a recent Flow switcher who didn't know that `?T` was illegal TS.

Reply:
Weird....